lib/AST/ASTImporter.cpp

```diff
@@ -7670,20 +7667,510 @@ Expected<QualType> ASTImporter::Import(QualType FromT) {
+        if (ExpectedSLoc L = Importer.Import(From.getProtocolLoc(I)))
+          To.setProtocolLoc(I, *L);
+        else
+          return L.takeError();
+      }
+    }
+    return Error::success();
+  }
+
+  Error VisitObjCObjectTypeLoc(ObjCObjectTypeLoc From) {
+    auto To = ToL.castAs<ObjCObjectTypeLoc>();
+
+    if (ExpectedSLoc L = Importer.Import(From.getTypeArgsLAngleLoc()))
+      To.setTypeArgsLAngleLoc(*L);
+    else
+      return L.takeError();
+    if (ExpectedSLoc L = Importer.Import(From.getTypeArgsRAngleLoc()))
+      To.setTypeArgsRAngleLoc(*L);
+    else
+      return L.takeError();
+
+    for (unsigned I = 0; I < From.getNumTypeArgs(); ++I) {
+      if (Expected<TypeSourceInfo *> TSI =
+              Importer.Import(From.getTypeArgTInfo(I)))
+        To.setTypeArgTInfo(I, *TSI);
+      else
+        return TSI.takeError();
+    }
+
+    if (ExpectedSLoc L = Importer.Import(From.getProtocolLAngleLoc()))
+      To.setProtocolLAngleLoc(*L);
+    else
+      return L.takeError();
+    if (ExpectedSLoc L = Importer.Import(From.getProtocolRAngleLoc()))
+      To.setProtocolRAngleLoc(*L);
+    else
+      return L.takeError();
+
+    for (unsigned I = 0; I < From.getNumProtocols(); ++I) {
+      if (ExpectedSLoc L = Importer.Import(From.getProtocolLoc(I)))
+        To.setProtocolLoc(I, *L);
+      else
+        return L.takeError();
+    }
+
+    To.setHasBaseTypeAsWritten(From.hasBaseTypeAsWritten());
+
+    return Error::success();
+  }
+
+  Error VisitObjCInterfaceTypeLoc(ObjCInterfaceTypeLoc From) {
+    auto To = ToL.castAs<ObjCInterfaceTypeLoc>();
+
+    if (ExpectedSLoc L = Importer.Import(From.getNameLoc()))
+      To.setNameLoc(*L);
+    else
+      return L.takeError();
+
+    return Error::success();
+  }
+
+  Error VisitObjCObjectPointerTypeLoc(ObjCObjectPointerTypeLoc From) {
+    auto To = ToL.castAs<ObjCObjectPointerTypeLoc>();
+    if (ExpectedSLoc L = Importer.Import(From.getStarLoc()))
+      To.setStarLoc(*L);
+    else
+      return L.takeError();
+    return Error::success();
+  }
+
+  Error VisitTypeLoc(TypeLoc TyLoc) { return Error::success(); }
+};
+
+} // namespace clang
+
 Expected<TypeSourceInfo *> ASTImporter::Import(TypeSourceInfo *FromTSI) {
   if (!FromTSI)
     return FromTSI;
 
-  // FIXME: For now we just create a "trivial" type source info based
-  // on the type and a single location. Implement a real version of this.
   ExpectedType TOrErr = Import(FromTSI->getType());
   if (!TOrErr)
     return TOrErr.takeError();
-  ExpectedSLoc BeginLocOrErr = Import(FromTSI->getTypeLoc().getBeginLoc());
-  if (!BeginLocOrErr)
-    return BeginLocOrErr.takeError();
 
-  return ToContext.getTrivialTypeSourceInfo(*TOrErr, *BeginLocOrErr);
+  if (Minimal) {
+    ExpectedSLoc BeginLocOrErr = Import(FromTSI->getTypeLoc().getBeginLoc());
+    if (!BeginLocOrErr)
+      return BeginLocOrErr.takeError();
+
+    return ToContext.getTrivialTypeSourceInfo(*TOrErr, *BeginLocOrErr);
```

Comment:
What is the exact difference of `getTrivialTypeSourceInfo` and between the new `TypeLocImporter` ?
Why is it okay to use the former in case of Minimal import?

Reply:
The trivial type source info includes only the one passed `SourceLocation` (the begin loc) in all places (where normally different locations are used). It is as like the `TypeLocImporter` would set all the locations to the one BeginLoc instead of importing these (if I observed the code correctly). At minimal import I did not want to import other non-related objects (`Decl`) that may come from import of `TypeSourceInfo` (if it refers to record) (and there is the FIXME case with the function parameters). (The minimal import should not cause other `Decl` objects to be imported that reside outside of the originally imported one?) It may be tried again without this `if`, probably there is no error now (it is a fix for some encountered problem).

Reply:
Okay, this makes sense for me.
> It may be tried again without this `if`
No, it looks correct the way as it is now,